sdk/appconfiguration/Azure.Data.AppConfiguration/src/AuthenticationPolicy.cs

```diff
@@ -23,62 +23,82 @@ public AuthenticationPolicy(string credential, byte[] secret)
             _secret = secret;
         }
 
-        public override async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        public override void Process(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
         {
-            await ProcessAsync(message, async: true).ConfigureAwait(false);
+            string contentHash = CreateContentHash(message);
+            AddHeaders(message, contentHash);
+            ProcessNext(message, pipeline);
+        }
 
+        public override async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            string contentHash = await CreateContentHashAsync(message).ConfigureAwait(false);
+            AddHeaders(message, contentHash);
             await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
         }
 
-        private async ValueTask ProcessAsync(HttpMessage message, bool async)
+        private static string CreateContentHash(HttpMessage message)
         {
-            string contentHash;
+            using var alg = SHA256.Create();
 
-            using (var alg = SHA256.Create())
+            using (var memoryStream = new MemoryStream())
+            using (var contentHashStream = new CryptoStream(memoryStream, alg, CryptoStreamMode.Write))
             {
-                using (var memoryStream = new MemoryStream())
-                using (var contentHashStream = new CryptoStream(memoryStream, alg, CryptoStreamMode.Write))
-                {
-                    if (message.Request.Content != null)
-                    {
-                        if (async)
-                        {
-                            await message.Request.Content.WriteToAsync(contentHashStream, message.CancellationToken).ConfigureAwait(false);
-                        }
-                        else
-                        {
-                            message.Request.Content.WriteTo(contentHashStream, message.CancellationToken);
-                        }
-                    }
-                }
-
-                contentHash = Convert.ToBase64String(alg.Hash);
+                message.Request.Content?.WriteTo(contentHashStream, message.CancellationToken);
             }
 
-            using (var hmac = new HMACSHA256(_secret))
+            return Convert.ToBase64String(alg.Hash);
+        }
+
+        private static async ValueTask<string> CreateContentHashAsync(HttpMessage message)
```

Comment:
Why split this into sync and async versions, rather than using the prior approach?

Reply:
Sync-over-async and async-over-sync code is always a bug prone.

In this particular case, we had (before the change) sync method `Process` that calls async method `ProcessAsync`, which calls private `ProcessAsync` with `async: false` but async `ProcessNextAsync`, so [`GetAwaiter().GetResult()`](https://github.com/Azure/azure-sdk-for-net/blob/ba2599979ea02654a2a6297c08c8b957aec4e10c/sdk/appconfiguration/Azure.Data.AppConfiguration/src/AuthenticationPolicy.cs#L79) can be called on the incomplete `ValueTask`, which is [not allowed](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask-1?view=netcore-3.0#remarks):

> The following operations should never be performed on a ValueTask<TResult> instance:
> - Awaiting the instance multiple times.
> - Calling AsTask multiple times.
> - Using .Result or .GetAwaiter().GetResult() when the operation hasn't yet completed, or using them multiple times.
> - Using more than one of these techniques to consume the instance.

There is also a classic sync context problem. User-defined `HttpPipelinePolicy` instances can be added by `ClientOptions`. If some WPF application adds a policy with UI-thread interaction, this code will hang.